utils/refactor-check-compiles.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+from __future__ import print_function
+
+import argparse
+import os
+import subprocess
+import sys
+
+
+def run_cmd(cmd, desc):
+    try:
+        return subprocess.check_output(cmd)
+    except subprocess.CalledProcessError:
+        print('FAILED ' + desc + ':', file=sys.stderr)
+        print(' '.join(cmd), file=sys.stderr)
+        sys.exit(1)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description="""
+        A drop-in replacement for a 'swift-refactor -dump-text' call that
+        1. Checkes that the file still compiles after the refactoring by doing
+           'swift-refactor -dump-rewritten' and feeding the result to
+           'swift-frontend -typecheck'
+        2. Outputting the result of the 'swift-refactor -dump-text' call
+
+        All arguments other than the following will be forwarded to :
+        'swift-refactor':
+         - swift-frontend
+         - swift-refactor
+         - temp-dir
+        """)
+
+    parser.add_argument(
+        '-swift-frontend',
+        help='The path to the swift-frontend executable'
+    )
+    parser.add_argument(
+        '-swift-refactor',
+        help='The path to the swift-refactor executable'
+    )
+    parser.add_argument(
+        '-temp-dir',
+        help='A temporary directory to save the rewritten file in'
+    )
+    parser.add_argument(
```

Comment:
Should we remove this altogether? It's a little weird because it isn't actually required. If we don't want `-dump-text` all the time we could have a `-dump-as` parameter to specify the output type. I realise this makes it not a drop-in replacement, but all those lines are changing anyway so it doesn't seem that bad to me.

I'd personally prefer to avoid the two swift-refactor calls. We could have a consumer that runs both rewritten and text?

Reply:
> Should we remove this altogether? It's a little weird because it isn't actually required. If we don't want `-dump-text` all the time we could have a `-dump-as` parameter to specify the output type. I realise this makes it not a drop-in replacement, but all those lines are changing anyway so it doesn't seem that bad to me.

The way I thought about it, was that the `-dump-text` makes it explicit what is printed to stdout by `%refactor-check-compiles`. Also could also one day add an option to `%refactor-check-compiles` that also supports `-dump-rewritten`. Plus, I like the idea of the drop-in replacement because it makes it super easy to switch between the two.

> I'd personally prefer to avoid the two swift-refactor calls. We could have a consumer that runs both rewritten and text?

I agree that it would be more performant to only call `swift-refactor` once but then that would need to separate the two outputs somehow, we’d need to split them again and it all becomes more fragile. Also, I don’t think it’s a huge performance bottleneck during testing right now. What do you think about adding a TODO comment about it and opening a bug on bugs.swift.org for it?

Reply:
> Also could also one day add an option to %refactor-check-compiles that also supports -dump-rewritten

Fair enough

> we’d need to split them again and it all becomes more fragile

Yeah, I was more thinking we'd just have `dump-rewritten` output to a file instead

> What do you think about adding a TODO comment about it and opening a bug on bugs.swift.org for it?

All good, it's fine as is for now

Reply:
Added a FIXME and filed https://bugs.swift.org/browse/SR-14587